### PR TITLE
Move fieldmanager

### PR DESCRIPTION
**Pull request by apelisse**
This is mostly a clean-up of the API of the fieldmanager package, there are more clean-ups that I think I can do now that I've removed some public things, but I'm going step-by-step. I think eventually I'll move this out of `apiserver/pkg/endpoints/handler` to put it in a place where it's more re-usable.

All the functions and types removed had no reasons to be public in the first place, and I don't really see a reason why they would be used by someone else.

Here's a diff of the old go-doc vs new go-doc, to make it clear how the API is affected:
```diff
--- old	2022-12-15 09:51:52.000000000 -0800
+++ new	2022-12-15 09:51:46.000000000 -0800
@@ -3,18 +3,6 @@
 
 CONSTANTS
 
-const DefaultMaxUpdateManagers int = 10
-    DefaultMaxUpdateManagers defines the default maximum retained number of
-    managedFields entries from updates if the number of update managers exceeds
-    this, the oldest entries will be merged until the number is below the
-    maximum. TODO(jennybuckley): Determine if this is really the best value.
-    Ideally we wouldn't unnecessarily merge too many entries.
-
-const DefaultTrackOnCreateProbability float32 = 1
-    DefaultTrackOnCreateProbability defines the default probability that the
-    field management of an object starts being tracked from the object's
-    creation, instead of from the first time the object is applied to.
-
 const InvalidManagedFieldsAfterMutatingAdmissionWarningFormat = ".metadata.managedFields was in an invalid state after admission; this could be caused by an outdated mutating admission controller; please fix your requests: %v"
     InvalidManagedFieldsAfterMutatingAdmissionWarningFormat is the warning that
     a client receives when a create/update/patch request results in invalid
@@ -39,27 +27,8 @@
 
 TYPES
 
-type DeducedTypeConverter struct{}
-    DeducedTypeConverter is a TypeConverter for CRDs that don't have a schema.
-    It does implement the same interface though (and create the same types of
-    objects), so that everything can still work the same. CRDs are merged with
-    all their fields being "atomic" (lists included).
-
-    Note that this is not going to be sufficient for converting to/from
-    CRDs that have a schema defined (we don't support that schema yet).
-    TODO(jennybuckley): Use the schema provided by a CRD if it exists.
-
-func (DeducedTypeConverter) ObjectToTyped(obj runtime.Object) (*typed.TypedValue, error)
-    ObjectToTyped converts an object into a TypedValue with a "deduced type".
-
-func (DeducedTypeConverter) TypedToObject(value *typed.TypedValue) (runtime.Object, error)
-    TypedToObject transforms the typed value into a runtime.Object. That is not
-    specific to deduced type.
-
-type FieldManager struct {
-	// Has unexported fields.
-}
-    FieldManager updates the managed fields and merge applied configurations.
+type FieldManager = internal.FieldManager
+    FieldManager updates the managed fields and merges applied configurations.
 
 func NewDefaultCRDFieldManager(typeConverter TypeConverter, objectConverter runtime.ObjectConvertor, objectDefaulter runtime.ObjectDefaulter, objectCreater runtime.ObjectCreater, kind schema.GroupVersionKind, hub schema.GroupVersion, subresource string, resetFields map[fieldpath.APIVersion]*fieldpath.Set) (_ *FieldManager, err error)
     NewDefaultCRDFieldManager creates a new FieldManager specifically for CRDs.
@@ -70,102 +39,6 @@
     NewDefaultFieldManager creates a new FieldManager that merges apply requests
     and update managed fields for other types of requests.
 
-func NewFieldManager(f Manager, subresource string) *FieldManager
-    NewFieldManager creates a new FieldManager that decodes, manages, then
-    re-encodes managedFields on update and apply requests.
-
-func (f *FieldManager) Apply(liveObj, appliedObj runtime.Object, manager string, force bool) (object runtime.Object, err error)
-    Apply is used when server-side apply is called, as it merges the object and
-    updates the managed fields.
-
-func (f *FieldManager) Update(liveObj, newObj runtime.Object, manager string) (object runtime.Object, err error)
-    Update is used when the object has already been merged (non-apply use-case),
-    and simply updates the managed fields in the output object.
-
-func (f *FieldManager) UpdateNoErrors(liveObj, newObj runtime.Object, manager string) runtime.Object
-    UpdateNoErrors is the same as Update, but it will not return errors.
-    If an error happens, the object is returned with managedFields cleared.
-
-type Managed interface {
-	// Fields gets the fieldpath.ManagedFields.
-	Fields() fieldpath.ManagedFields
-
-	// Times gets the timestamps associated with each operation.
-	Times() map[string]*metav1.Time
-}
-    Managed groups a fieldpath.ManagedFields together with the timestamps
-    associated with each operation.
-
-func DecodeManagedFields(encodedManagedFields []metav1.ManagedFieldsEntry) (Managed, error)
-    DecodeManagedFields converts ManagedFields from the wire format (api format)
-    to the format used by sigs.k8s.io/structured-merge-diff
-
-type Manager interface {
-	// Update is used when the object has already been merged (non-apply
-	// use-case), and simply updates the managed fields in the output
-	// object.
-	//  * `liveObj` is not mutated by this function
-	//  * `newObj` may be mutated by this function
-	// Returns the new object with managedFields removed, and the object's new
-	// proposed managedFields separately.
-	Update(liveObj, newObj runtime.Object, managed Managed, manager string) (runtime.Object, Managed, error)
-
-	// Apply is used when server-side apply is called, as it merges the
-	// object and updates the managed fields.
-	//  * `liveObj` is not mutated by this function
-	//  * `newObj` may be mutated by this function
-	// Returns the new object with managedFields removed, and the object's new
-	// proposed managedFields separately.
-	Apply(liveObj, appliedObj runtime.Object, managed Managed, fieldManager string, force bool) (runtime.Object, Managed, error)
-}
-    Manager updates the managed fields and merges applied configurations.
-
-func NewBuildManagerInfoManager(f Manager, gv schema.GroupVersion, subresource string) Manager
-    NewBuildManagerInfoManager creates a new Manager that converts the manager
-    name into a unique identifier combining operation and version for update
-    requests, and just operation for apply requests.
-
-func NewCRDStructuredMergeManager(typeConverter TypeConverter, objectConverter runtime.ObjectConvertor, objectDefaulter runtime.ObjectDefaulter, gv schema.GroupVersion, hub schema.GroupVersion, resetFields map[fieldpath.APIVersion]*fieldpath.Set) (_ Manager, err error)
-    NewCRDStructuredMergeManager creates a new Manager specifically for CRDs.
-    This allows for the possibility of fields which are not defined in models,
-    as well as having no models defined at all.
-
-func NewCapManagersManager(fieldManager Manager, maxUpdateManagers int) Manager
-    NewCapManagersManager creates a new wrapped FieldManager which ensures that
-    the number of managers from updates does not exceed maxUpdateManagers,
-    by merging some of the oldest entries on each update.
-
-func NewLastAppliedManager(fieldManager Manager, typeConverter TypeConverter, objectConverter runtime.ObjectConvertor, groupVersion schema.GroupVersion) Manager
-    NewLastAppliedManager converts the client-side apply annotation to
-    server-side apply managed fields
-
-func NewLastAppliedUpdater(fieldManager Manager) Manager
-    NewLastAppliedUpdater sets the client-side apply annotation up to date with
-    server-side apply managed fields
-
-func NewManagedFieldsUpdater(fieldManager Manager) Manager
-    NewManagedFieldsUpdater is responsible for updating the managedfields in
-    the object, updating the time of the operation as necessary. For updates,
-    it uses a hard-coded manager to detect if things have changed, and swaps
-    back the correct manager after the operation is done.
-
-func NewProbabilisticSkipNonAppliedManager(fieldManager Manager, objectCreater runtime.ObjectCreater, gvk schema.GroupVersionKind, p float32) Manager
-    NewProbabilisticSkipNonAppliedManager creates a new wrapped FieldManager
-    that starts tracking managers after the first apply, or starts tracking on
-    create with p probability.
-
-func NewSkipNonAppliedManager(fieldManager Manager, objectCreater runtime.ObjectCreater, gvk schema.GroupVersionKind) Manager
-    NewSkipNonAppliedManager creates a new wrapped FieldManager that only starts
-    tracking managers after the first apply.
-
-func NewStripMetaManager(fieldManager Manager) Manager
-    NewStripMetaManager creates a new Manager that strips metadata and typemeta
-    fields from the manager's fieldset.
-
-func NewStructuredMergeManager(typeConverter TypeConverter, objectConverter runtime.ObjectConvertor, objectDefaulter runtime.ObjectDefaulter, gv schema.GroupVersion, hub schema.GroupVersion, resetFields map[fieldpath.APIVersion]*fieldpath.Set) (Manager, error)
-    NewStructuredMergeManager creates a new Manager that merges apply requests
-    and update managed fields for other types of requests.
-
 type ResourcePathMappings map[string]fieldpath.Path
     ResourcePathMappings maps a group/version to its replicas path. The
     assumption is that all the paths correspond to leaf fields.
@@ -191,13 +64,20 @@
      2. Replicas path of the main resource is transformed to the replicas path
         of the scale subresource
 
-type TypeConverter interface {
-	ObjectToTyped(runtime.Object) (*typed.TypedValue, error)
-	TypedToObject(*typed.TypedValue) (runtime.Object, error)
-}
+type TypeConverter = internal.TypeConverter
     TypeConverter allows you to convert from runtime.Object to typed.TypedValue
     and the other way around.
 
+func NewDeducedTypeConverter() TypeConverter
+    NewDeducedTypeConverter creates a TypeConverter for CRDs that don't have
+    a schema. It does implement the same interface though (and create the same
+    types of objects), so that everything can still work the same. CRDs are
+    merged with all their fields being "atomic" (lists included).
+
+    Note that this is not going to be sufficient for converting to/from
+    CRDs that have a schema defined (we don't support that schema yet).
+    TODO(jennybuckley): Use the schema provided by a CRD if it exists.
+
 func NewTypeConverter(models proto.Models, preserveUnknownFields bool) (TypeConverter, error)
     NewTypeConverter builds a TypeConverter from a proto.Models. This will
     automatically find the proper version of the object, and the corresponding
```

#### What type of PR is this?
/kind cleanup

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/versionconverter.go [hunk 1]
```diff
@@ -99,3 +101,23 @@ func (v *versionConverter) Convert(object *typed.TypedValue, version fieldpath.A
 func (v *versionConverter) IsMissingVersionError(err error) bool {
 	return runtime.IsNotRegisteredError(err) || isNoCorrespondingTypeError(err)
 }
+
+type noCorrespondingTypeErr struct {
```
Comment (alexzielenski) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/versionconverter.go:105]: Does it make sense for this error type to be private? Given that it is returned by a public function? That makes it impossible to use `errors.As` correct?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/typeconverter.go [hunk 2]
```diff
@@ -17,114 +17,30 @@ limitations under the License.
...
-// CRDs are merged with all their fields being "atomic" (lists
+// NewDeducedTypeConverter creates a TypeConverter for CRDs that don't
+// have a schema. It does implement the same interface though (and
+// create the same types of objects), so that everything can still work
+// the same. CRDs are merged with all their fields being "atomic" (lists
 // included).
 //
 // Note that this is not going to be sufficient for converting to/from
 // CRDs that have a schema defined (we don't support that schema yet).
 // TODO(jennybuckley): Use the schema provided by a CRD if it exists.
```
Comment (alexzielenski) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/typeconverter.go:36]: comment can be removed too

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanagertest/testfieldmanager.go [hunk 3]
```diff
@@ -128,10 +129,10 @@ func NewDefaultTestFieldManager(typeConverter fieldmanager.TypeConverter, gvk sc
 }
 
 // NewTestFieldManager creates a new manager for the given GVK.
-func NewTestFieldManager(typeConverter fieldmanager.TypeConverter, gvk schema.GroupVersionKind, subresource string, chainFieldManager func(fieldmanager.Manager) fieldmanager.Manager) TestFieldManager {
+func NewTestFieldManager(typeConverter fieldmanager.TypeConverter, gvk schema.GroupVersionKind, subresource string, chainFieldManager func(internal.Manager) internal.Manager) TestFieldManager {
```
Comment (alexzielenski) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanagertest/testfieldmanager.go:132]: Surprising to me that this public(?) api has an argument that uses internal types. Is the test field manager expected to be used outside of this package?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/fieldmanager.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
+/*
```
Comment (alexzielenski) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/fieldmanager.go:1]: It's hard to tell what changes were made, if any, to the methods in this file, but it seems ok on surface

**Review by alexzielenski — CHANGES_REQUESTED** [review]
some minor questions

looks clean. I like the reduced API Surface area

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanagertest/testfieldmanager.go (hunk 3 above)
Comment (apelisse) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanagertest/testfieldmanager.go:132]: I don't even know how the compiler can let me do this yeah :-)
This was the next function I wanted to clean, maybe I should clean that before we get it merged.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/fieldmanager.go (hunk 4 above)
Comment (apelisse) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/fieldmanager.go:1]: Yeah, there are no changes beside `fieldmanager.` -> `internal.` AFAIR

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/versionconverter.go (hunk 1 above)
Comment (apelisse) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/versionconverter.go:105]: Note that this was private already, this is using the pattern described in this articl: https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully
This is mostly unrelated to the change, I just moved this.
/shrug

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/typeconverter.go (hunk 2 above)
Comment (apelisse) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/typeconverter.go:36]: yeah, I'll clean that up, it's unrelated to the change but I'll push into another commit

**Comment by apelisse** [comment]
OK I've now also reduced the api surface of the test package, here's the diff:
```diff
--- old	2022-12-15 20:51:15.000000000 -0800
+++ new	2022-12-15 20:51:28.000000000 -0800
@@ -3,19 +3,29 @@
 
 FUNCTIONS
 
-func NewBuiltinTypeConverter() fieldmanager.TypeConverter
+func NewTypeConverter(schema prototesting.Fake) fieldmanager.TypeConverter
     NewBuiltinTypeConverter creates a TypeConverter with all the built-in types
     defined, given the committed kubernetes swagger.json.
 
-func NewFakeObjectCreater() runtime.ObjectCreater
-    NewFakeObjectCreater implements ObjectCreater, it can create empty objects
-    (unstructured) of the given GVK.
-
 
 TYPES
 
-type TestFieldManager struct {
-	// Has unexported fields.
+type TestFieldManager interface {
+	// APIVersion of the object that we're tracking.
+	APIVersion() string
+	// Reset resets the state of the liveObject by resetting it to an empty object.
+	Reset()
+	// Live returns a copy of the current liveObject.
+	Live() runtime.Object
+	// Apply applies the given object on top of the current liveObj, for the
+	// given manager and force flag.
+	Apply(obj runtime.Object, manager string, force bool) error
+	// Update will updates the managed fields in the liveObj based on the
+	// changes performed by the update.
+	Update(obj runtime.Object, manager string) error
+	// ManagedFields returns the list of existing managed fields for the
+	// liveObj.
+	ManagedFields() []metav1.ManagedFieldsEntry
 }
     TestFieldManager is a FieldManager that can be used in test to simulate
     the behavior of Server-Side Apply and field tracking. This also has a few
@@ -29,30 +39,11 @@
     specify either a sub-resource, or a set of modified Manager to test them
     specifically.
 
-func NewDefaultTestFieldManager(gvk schema.GroupVersionKind) TestFieldManager
+func NewTestFieldManager(typeConverter fieldmanager.TypeConverter, gvk schema.GroupVersionKind) TestFieldManager
     NewDefaultTestFieldManager returns a new TestFieldManager built for the
     given gvk, on the main resource.
 
-func NewTestFieldManager(gvk schema.GroupVersionKind, subresource string, chainFieldManager func(fieldmanager.Manager) fieldmanager.Manager) TestFieldManager
-    NewTestFieldManager creates a new manager for the given GVK.
-
-func (f *TestFieldManager) APIVersion() string
-    APIVersion of the object that we're tracking.
-
-func (f *TestFieldManager) Apply(obj runtime.Object, manager string, force bool) error
-    Apply applies the given object on top of the current liveObj, for the given
-    manager and force flag.
-
-func (f *TestFieldManager) Live() runtime.Object
-    Live returns a copy of the current liveObject.
-
-func (f *TestFieldManager) ManagedFields() []metav1.ManagedFieldsEntry
-    ManagedFields returns the list of existing managed fields for the liveObj.
-
-func (f *TestFieldManager) Reset()
-    Reset resets the state of the liveObject by resetting it to an empty object.
-
-func (f *TestFieldManager) Update(obj runtime.Object, manager string) error
-    Update will updates the managed fields in the liveObj based on the changes
-    performed by the update.
+func NewTestFieldManagerSubresource(typeConverter fieldmanager.TypeConverter, gvk schema.GroupVersionKind, subresource string) TestFieldManager
+    NewDefaultTestFieldManager returns a new TestFieldManager built for the
+    given gvk, on the main resource.
 ```

That's not much smaller, but that's definitely cleaner.

**Comment by apelisse** [comment]
I think I can further simplify the "NewTypeConverter" API to just take a schema, which doesn't have to be different for the test version than the actual verison, then when we swap the typeconverter for a direct conversion (i.e. with Alex's PR), it should be entirely transparent.

**Comment by apelisse** [comment]
Replaced with https://github.com/kubernetes/kubernetes/pull/114977

**Pull request closed without merging**